### PR TITLE
Inet implementations assume BYTE_ORDER is defined. make sure that it is

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
BYTE_ORDER was not defined in Silabs Thread apps. But the inet implementation assumes it is for some endianness conversion.

Recent changes to GCC highlighted this missing define as we were now using the big-endian implementation of `hotnl` on the Silabs platform due to `#if BYTE_ORDER == BIG_ENDIAN` evaluating to true when both of the defines didn't exist.
This, ultimately, led to `IPAddress::IPAddress::MakeIPv6Multicast` returning an invalid Multicast address.

Anyhow, this PR adds some checks on the existence of the BYTE_ORDER defined.

for `CHIP_SYSTEM_CONFIG_USE_OPEN_THREAD_ENDPOINT` inet implementation, Try at least to define BYTE_ORDER to the right endianness of the system if the definition doesn't exist already.

Note: Platforms using `lwip` usually define `BYTE_ORDER`, `BIG_ENDIAN`, `LITTLE_ENDIAN` in `arch/cc.h`

### CODE REVIEW

File: src/inet/arpa-inet-compatibility.h [hunk 1]
```diff
@@ -49,6 +56,26 @@
 
 #if CHIP_SYSTEM_CONFIG_USE_OPEN_THREAD_ENDPOINT
 
+#ifndef BYTE_ORDER
+#if defined(__LITTLE_ENDIAN__)
+#define BYTE_ORDER LITTLE_ENDIAN
+#elif defined(__BIG_ENDIAN__)
+#define BYTE_ORDER BIG_ENDIAN
+#elif defined(__BYTE_ORDER__)
+#define BYTE_ORDER __BYTE_ORDER__
```
Comment (andy31415) [inline src/inet/arpa-inet-compatibility.h:65]: you need to define BIG_ENDIAN and LITTLE_ENDIAN as well then.

File: src/inet/arpa-inet-compatibility.h [hunk 2]
```diff
@@ -49,6 +56,26 @@
 
 #if CHIP_SYSTEM_CONFIG_USE_OPEN_THREAD_ENDPOINT
 
+#ifndef BYTE_ORDER
+#if defined(__LITTLE_ENDIAN__)
+#define BYTE_ORDER LITTLE_ENDIAN
```
Comment (andy31415) [inline src/inet/arpa-inet-compatibility.h:61]: This does not work if LITTLE_ENDIAN is not defined right? In the new GCC case, neither LITTLE_ENDIAN nor BIG_ENDIAN nor BYTE_ORDER are defined.

lets not try to use BYTE_ORDER at all ... it does not seem standard.

Comment (jmartinez-silabs) [reply src/inet/arpa-inet-compatibility.h:61]: Quote what I mentioned on slack

> from what I understood, different scenarios with different possibles macros
> In the case we're __LITTLE_ENDIAN__ or __BIG_ENDIAN__ their counterparts without then__ prefix/suffix will also be defined with a set value
> In the case they are not and __BYTE_ORDER__ is defined then __ORDER_LITTLE_ENDIAN__  should also be defined. That's why I do #define LITTLE_ENDIAN __ORDER_LITTLE_ENDIAN__ after.

But for simplicity ill remove the other condition and BYTE_ORDER all together and only use `__BYTE_ORDER__`

If there is a case or platform that do not have this define then at least the #ifndef error will prevent the build and notify the developer (and CI should catch it)

File: src/inet/arpa-inet-compatibility.h [hunk 3]
```diff
@@ -49,7 +48,11 @@
 
 #if CHIP_SYSTEM_CONFIG_USE_OPEN_THREAD_ENDPOINT
 
-#if BYTE_ORDER == BIG_ENDIAN
+#ifndef __BYTE_ORDER__
+#error Endianness not defined is not defined
```
Comment (bzbarsky-apple) [inline src/inet/arpa-inet-compatibility.h:52]:
```suggestion
#error Endianness is not defined
```

File: src/inet/arpa-inet-compatibility.h [hunk 4]
```diff
@@ -78,7 +81,10 @@
 #ifndef ntohl
 #define ntohl(x) htonl(x)
 #endif
-#endif // BYTE_ORDER == BIG_ENDIAN
+
+#else
+#error Unknown endiannes set to _BYTE_ORDER__
```
Comment (bzbarsky-apple) [inline src/inet/arpa-inet-compatibility.h:86]:
```suggestion
#error __BYTE_ORDER__ value not recognized
```

File: src/inet/arpa-inet-compatibility.h [hunk 5]
```diff
@@ -78,7 +81,10 @@
...
+#endif // _BYTE_ORDER__ ==
```
Comment (bzbarsky-apple) [inline src/inet/arpa-inet-compatibility.h:87]:
```suggestion
#endif // __BYTE_ORDER__ ==
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by arkq — APPROVED** [review]

**Pull request merged**